yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala

```diff
@@ -82,6 +100,179 @@ class YarnSparkHadoopUtil extends SparkHadoopUtil {
     if (credentials != null) credentials.getSecretKey(new Text(key)) else null
   }
 
+  private[spark] override def scheduleLoginFromKeytab(): Unit = {
+    sparkConf.getOption("spark.yarn.principal").foreach { principal =>
+      val keytab = sparkConf.get("spark.yarn.keytab")
+      val delegationTokenRenewerRunnable =
```

Comment:
Wait, I'm a little confused.

`delegationTokenExecuterUpdaterRunnable` runs on the driver, right? And `delegationTokenRenewerRunnable` runs on the executors, right?

Could you add comments to these (and maybe tweak the variable names) so that this is clearer? The names are very similar and it's easy to miss the difference here.

Also, you could stash this in a lazy val like the other runnable.